allennlp/training/metrics/rouge.py

```diff
@@ -0,0 +1,120 @@
+from collections import Counter
+from typing import Tuple, Dict, Set
+
+from overrides import overrides
+import torch
+
+from allennlp.training.metrics.metric import Metric
+
+
+# TODO: implement ROUGE-L
+
+
+@Metric.register("rogue")
+class ROUGE(Metric):
+    """
+    Recall-Oriented Understudy for Gisting Evaluation (ROUGE)
+
+    ROUGE is a metric for measuring the quality of summaries. It is based on calculating the recall
+    between ngrams in the predicted summary and a set of reference summaries. See [Lin,
+    "ROUGE: A Package For Automatic Evaluation Of Summaries", 2004]
+    (https://api.semanticscholar.org/CorpusID:964287).
+
+    # Parameters
+
+    ngram_size : `int`, optional (default = 2)
+        ROUGE scores are calculate for ROUGE-1 .. ROUGE-`ngram_size`
+    exclude_indices : `Set[int]`, optional (default = None)
+        Indices to exclude when calculating ngrams. This should usually include
+        the indices of the start, end, and pad tokens.
+    """
+
+    def __init__(self, ngram_size: int = 2, exclude_indices: Set[int] = None,) -> None:
+        self._ngram_size = ngram_size
+        self._exclude_indices = exclude_indices or set()
+        self._recall_matches: Dict[int, int] = Counter()
+        self._recall_totals: Dict[int, int] = Counter()
+
+    @overrides
+    def reset(self) -> None:
+        self._recall_matches = Counter()
+        self._recall_totals = Counter()
+
+    def _get_recall_counts(
+        self,
+        predicted_tokens: torch.LongTensor,
+        reference_tokens: torch.LongTensor,
+        ngram_size: int,
+    ) -> Tuple[int, int]:
+        """
+        Compare the predicted tokens to the reference (gold) tokens at the desired
+        ngram size and calculate the numerator and denominator for recall.
+
+        The numerator is the number of ngrams in the predicted sentences that match
+        with an ngram in the corresponding reference sentence, clipped by the total
+        count of that ngram in the reference sentence. The denominator is
+        the total count of reference ngrams.
+        """
+        # TODO: fix not being able to import this normally
+        from allennlp.training.util import ngrams
+
+        matches = 0
+        total_reference = 0
+        for batch_num in range(predicted_tokens.size(0)):
+            predicted_row = predicted_tokens[batch_num, :]
+            reference_row = reference_tokens[batch_num, :]
```

Comment:
You could just say `for predicted_row, reference_row in zip(predicted_tokens, reference_tokens):`.

Reply:
Thanks that looks better. I took this code from the bleu class, so I changed it in rouge and bleu.